### ISSUE TITLE
Images are not being correctly inserted

### ISSUE THREAD

**Issue by abelsromero** [asciidoctor/asciidoctor-pdf#87]
Images which are in the images directory are not included in the final rendered document.
During render the following message is shown:
```
asciidoctor: WARNING: could not embed image; images/${LOCAL_PATH}/asciidoctor-maven-examples/asciidoctor-pdf-example/src/docs/asciidoc/images/sunset.jpg n
ot found
```

Seems like the path is incorrectly appended at the beginning of the full path.
As an example you can test it with `asciidoctor-pdf-example` in [asciidoctor-maven-examples](https://github.com/asciidoctor/asciidoctor-maven-examples).

This has been tested setting the `images` parameter in the pom.xml to other values and the behaviour is the same, the parameter's value is appended at the beginning.

Related to asciidoctor/asciidoctor-maven-examples#11

**Comment by mojavelinux** [comment]
Aha! I see a note in the code warning me that this might happen.

See https://github.com/asciidoctor/asciidoctor-pdf/blob/master/lib/asciidoctor-pdf/converter.rb#L575-L576

**Comment by DavidGamba** [comment]
I was getting the same error when using the ruby API, not the command line:

asciidoctor: WARNING: could not embed image; /images/resmd/mandated_access.png not found

That is because `node.attr 'docdir'` [here](https://github.com/asciidoctor/asciidoctor-pdf/blob/02728c36bdaa04fb1d1036e425e6dd8e5a60dc9b/lib/asciidoctor-pdf/converter.rb#L575-L576) returns an empty string and I get an absolute path as a result:

````
irb
2.2.0 :001 > require 'fileutils'
 => true 
2.2.0 :002 > File.join('',"test")
 => "/test" 
2.2.0 :003 > 
````
I initially tried removing the starting `/` by checking if `docdir` is an empty string:

````ruby
image_path = File.join((node.attr 'docdir'), (node.attr 'imagesdir') || '', target)
image_path[0] = '' if (node.attr 'docdir') == ''
````
but got this error:

asciidoctor: WARNING: could not embed image; images/images/resmd/mandated_access.png not found

So basically, when it finds a relative path it inserts the `imagesdir` again (haven't found where).

So  my fix for now is:

````diff
diff --git lib/asciidoctor-pdf/converter.rb lib/asciidoctor-pdf/converter.rb
index ab7f561..299712b 100644
--- lib/asciidoctor-pdf/converter.rb
+++ lib/asciidoctor-pdf/converter.rb
@@ -574,7 +574,10 @@ class Converter < ::Prawn::Document
     #end
 
     # FIXME use normalize_path here!
-    image_path = File.join((node.attr 'docdir'), (node.attr 'imagesdir') || '', target)
+    image_path = target
+
     # TODO extension should be an attribute on an image node
     image_type = File.extname(image_path)[1..-1]
     width = if node.attr? 'scaledwidth'
````

@mojavelinux WDYT?

**Comment by mojavelinux** [comment]
I believe I've fixed it. I'm still going to do some additional testing, including testing it against the Maven examples.

**Comment by mojavelinux** [comment]
Image resolution should be much more solid now. I still need to verify with the Maven plugin, but I think we've got it. I also added support for remote images :)

**Comment by mojavelinux** [comment]
I checked with the Maven plugin and it seems to resolve the path to the image correctly in all cases now.

**Comment by mojavelinux** [comment]
> I was getting the same error when using the ruby API

Make sure you set the safe mode to `:safe` or `:unsafe`. This fix isn't dependent on that setting, but it's recommended when using the API (the default is `:secure` which is overly conservative).

**Comment by Omnipresent** [comment]
I'm using 1.5.0

``` bash
▶ asciidoctor-pdf --version
Asciidoctor PDF 1.5.0.alpha.11 using Asciidoctor 1.5.4 [http://asciidoctor.org]
Runtime Environment (ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]) (lc:UTF-8 fs:UTF-8 in:- ex:UTF-8)
```
and when I convert the `adoc` into PDF, none of the images in my `images` folder are showing up. Everything works fine when I convert to HTML.

**Comment by mojavelinux** [comment]
@Omnipresent You'll need to be more specific or provide an example project.

**Comment by aalmiray** [comment]
I've got the same problem when adding `asciidoctorj-pdf:1.5.0-alpha.11` to the build https://github.com/griffon/griffon/blob/development/docs/griffon-guide/griffon-guide.gradle and enabling `pdf` as a backend (besides `html5`). 

    asciidoctor: WARNING: image to embed not found or not readable: /Users/aalmiray/dev/github/griffon/docs/griffon-guide/src/docs/asciidoc/images/griffon-banner-large.png
    asciidoctor: WARNING: image to embed not found or not readable: /Users/aalmiray/dev/github/griffon/docs/griffon-guide/src/docs/asciidoc/images/console-swing-groovy.png
    asciidoctor: WARNING: image to embed not found or not readable: /Users/aalmiray/dev/github/griffon/docs/griffon-guide/src/docs/asciidoc/images/sample-swing-java.png
    asciidoctor: WARNING: image to embed not found or not readable: /Users/aalmiray/dev/github/griffon/docs/griffon-guide/src/docs/asciidoc/images/sample-javafx-java.png
    asciidoctor: WARNING: image to embed not found or not readable: /Users/aalmiray/dev/github/griffon/docs/griffon-guide/src/docs/asciidoc/images/sample-lanterna-java.png
    asciidoctor: WARNING: image to embed not found or not readable: /Users/aalmiray/dev/github/griffon/docs/griffon-guide/src/docs/asciidoc/images/sample-pivot-java.png

those images are found under `src/resources` but the pdf expects them to be under `src/docs/asciidoc/`. Here's the full configuration for the asciidoctor task

    asciidoctor {
        backends   = ['html5', 'pdf']
        attributes = [
            toc                     : 'left',
            doctype                 : 'book',
            icons                   : 'font',
            encoding                : 'utf-8',
            sectlink                : true,
            sectanchors             : true,
            numbered                : true,
            linkattrs               : true,
            imagesdir               : 'images',
            linkcss                 : true,
            stylesheet              : 'css/style.css',
            'source-highlighter'    : 'coderay',
            'coderay-linenums-mode' : 'table',
            'griffon-version'       : project.version,
            'griffon-group'         : project.group,
            'griffon-source-url'    : griffonSourceUrl,
            'gradle-griffon-version': loadProperties(project(':gradle-griffon-plugin').file('gradle.properties')).version,
            'application-master-pom': loadProperties(project(':application-master-pom').file('gradle.properties')).version
        ]

        sources {
            include 'index.adoc'
        }

        resources {
            from file('src/resources')
            from(rootProject.file('src/media/banners/large.png')) {
                into 'images'
                rename '.*', 'griffon-banner-large.png'
            }
        }
    }

It's worth mentioning that the html5 backend works as expected.

**Comment by mojavelinux** [comment]
> It's worth mentioning that the html5 backend works as expected.

This boils down to who is resolving the images. In the case of the HTML converter, the viewer/browser resolves the images, so as long as they are where the `<img>` element says they are, everything works great. In the case of the PDF converter, it's the PDF converter that has to resolve the image. Therefore, the image is going to be read from a different location.

It's this reason why the imagesdir must always be set differently for different converters...and also why we have this abstraction in the first place. It's very unlikely the same value is going to work for all converters in the context of a build system (again, because different converters reference the files in different ways at different times).
